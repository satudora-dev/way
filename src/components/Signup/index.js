import React, { Component } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import EXIF from 'exif-js';
import SiteInfo from "../../components/SiteInfo";
import IconPreview from '../IconPreview';

const createObjectURL
  = (window.URL || window.webkitURL).createObjectURL || window.createObjectURL;

class Signup extends Component {
  constructor(props) {
    super(props);
    this.state = {
      givenName: "",
      familyName: "",
      sei: "",
      mei: "",
      iconFile: "",
      iconSrc: "/portrait.png",
      id: "",
    };

    this.onTextChange = this.onTextChange.bind(this);
  }


  onTextChange(e) {
    switch (e.target.name) {
      case 'givenName':
        this.setState({
          givenName: e.target.value,
        });
        break;
      case 'familyName':
        this.setState({
          familyName: e.target.value,
        });
        break;
      case 'sei':
        this.setState({
          sei: e.target.value,
        });
        break;
      case 'mei':
        this.setState({
          mei: e.target.value,
        });
        break;
    }
  }

  getIconFile(imageFile) {
    this.setState({
      iconFile: imageFile,
      iconSrc: createObjectURL(imageFile),
    });
  }

  render() {

    const style = {
      imagestyle: {
        width: 64,
        height: 32,
        "padding-top": "20px",
      },
      siteInfoStyle: {
        "margin-top": "30px",
      },
      welcomestyle: {
        color: "grey",
        "font-weight": "lighter",
      },
      pstyle: {
        "margin-bottom": "0px",
      },
      iconstyle: {
        width: 200,
        height: 200,
        "border-radius": "50%",
        "object-fit": "cover",
      },
      WAYstyle: {
        color: "black",
        "font-weight": "bold",
      }
    }

    return (
      <div className="Login">
        <div style={style.siteInfoStyle}>
          <SiteInfo />
        </div>
        <h3 style={style.welcomestyle}>Welcome.
        <span style={style.WAYstyle}> W</span>
          ho
        <span style={style.WAYstyle}> A</span>
          re
        <span style={style.WAYstyle}> Y</span>
          ou?
          </h3>
        <form noValidate autoComplete="off">
          <TextField
            name="givenName"
            label="given name"
            required={true}
            value={this.state.given_name}
            onChange={this.onTextChange}
            margin="normal"
            style={style.welcomestyle} />
          <TextField
            name="familyName"
            label="family name"
            required={true}
            value={this.state.family_name}
            onChange={this.onTextChange}
            margin="normal"
            style={style.welcomestyle} />
        </form>
        <form noValidate autoComplete="off">
          <TextField
            name="mei"
            label="名"
            required={true}
            value={this.state.mei}
            onChange={this.onTextChange}
            margin="normal"
            style={style.welcomestyle} />
          <TextField
            name="sei"
            label="姓"
            required={true}
            value={this.state.sei}
            onChange={this.onTextChange}
            margin="normal"
            style={style.welcomestyle} />
        </form>
        <IconPreview
          iconSrc={this.state.iconSrc}
          canEdit={true}
          onChange={imageFile =>
            this.getIconFile(imageFile)} />
        <Button variant="contained" onClick={() => {
          this.props.signUpAsUser(this.props.ownKey,
            this.state.givenName,
            this.state.familyName,
            this.state.mei,
            this.state.sei,
            this.state.iconFile);
        }}>
          GO
        </Button>
      </div>
    );
  }
}

export default Signup;
