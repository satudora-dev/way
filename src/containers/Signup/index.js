import React, { Component } from 'react';
import {firebaseAuth,firebaseDB,firebaseStorage} from '../../firebase';
import {withRouter,Link} from 'react-router-dom';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import CheckInstax from '../../components/CheckInstax';

class Signup extends Component {
  constructor(props){
    super(props);
    this.state={
      givenName: "",
      familyName: "",
      sei: "",
      mei: "",
      iconFile: "",
      iconSrc: "/portrait.png",
      id: "",
      onCheck: true,
    };

    this.onTextChange=this.onTextChange.bind(this);
    this.onSendProfile=this.onSendProfile.bind(this);
  }

  componentWillMount(){
    firebaseAuth().onAuthStateChanged(user=>{
      if(!user){
        this.props.history.push('/login');
        return;
      }

      let dbRef=firebaseDB.ref('accounts');
      dbRef.orderByChild('email').equalTo(user.email)//メールアドレスが既に登録されているか
        .once('value',(snapshot)=>{
          if(snapshot.val()==null){
            this.props.history.push('/login');
          }
          else{
            let registered;
            snapshot.forEach((childSnapshot)=>{
              registered=childSnapshot.child('registered').val();
              this.setState({id: childSnapshot.key});
            });
  
            if(registered){
              this.props.history.push('/users');
            }
            else{
              this.setState({onCheck: false});
            }
          }
      });
    });
  }

  onTextChange(e){
    switch(e.target.name){
      case 'givenName':
        this.setState({
          givenName: e.target.value,
        });
        break;
      case 'familyName':
        this.setState({
          familyName: e.target.value,
        });
        break;
      case 'sei':
        this.setState({
          sei: e.target.value,
        });
        break;
      case 'mei':
        this.setState({
          mei: e.target.value,
        });
        break;
      case 'icon':
        if(!e.target.files[0])return;
        this.resizeImage(e.target.files[0]);
        break;
    }
  }

  resizeImage(iconFile){
    let image=new Image();
    image.onload=()=>{
      let width=image.width;
      let height=image.height;
      let maxWidth=512;
      if(width<maxWidth){
        this.setState({
          iconFile: iconFile,
          iconSrc: URL.createObjectURL(iconFile),
        });
        return;
      }
      else{
        let scale=maxWidth/width;
        let dstHeight=height*scale;
        let canvas=document.createElement('canvas');
        canvas.width=maxWidth;
        canvas.height=dstHeight;
        let ctx=canvas.getContext('2d');
        ctx.drawImage(image,0,0,maxWidth,dstHeight);

        let resizedImage=canvas.toDataURL('image/png');
        canvas.toBlob((blob)=>{
          this.setState({iconFile: blob});
        });
        this.setState({iconSrc: resizedImage});
      }
    };
    image.src=URL.createObjectURL(iconFile);
  }

  onSendProfile(e){
    if(!this.state.givenName){
      alert('Given name is empty.');
      return;
    }
    else if(!this.state.familyName){
      alert('Famili name is empty.');
      return;
    }
    else if(!this.state.sei){
      alert('姓が未入力です。');
      return;
    }
    else if(!this.state.mei){
      alert('名が未入力です。');
      return;
    }
    else if(!this.state.iconFile){
      alert('Icon image is empty.');
      return;
    }
    //alert(this.state.iconFile);
    firebaseDB.ref('users/'+this.state.id).set({
      "given": this.state.givenName,
      "family": this.state.familyName,
      "sei": this.state.sei,
      "mei": this.state.mei,
      "haveIcon": this.state.iconFile!==""
    });
    firebaseDB.ref('accounts/'+this.state.id).update({'registered': true});
    
    if(this.state.iconFile){
      let storageRef=firebaseStorage.ref().child(
        'icons/'+this.state.id);
        storageRef.put(this.state.iconFile).then((snapshot)=>{
          this.props.history.push('/users');
      });
    }
    else{
      this.props.history.push('/users');
    }
  }

  render() {

    if(this.state.onCheck) return(<div></div>);

    const style = {
      imagestyle: {
        width: 64,
        height: 32,
        "padding-top": "20px",
      },
      welcomestyle: {
        color: "grey",
        "font-weight": "lighter",
      },
      pstyle: {
        "margin-bottom": "0px",
      },
      iconstyle: {
        width: 200,
        height:200,
        "border-radius": "50%",
        "object-fit": "cover",
      },
    }

    return (
      <div className="Login">
        <div>
          <Link to="/">
            <img src="way.png" style={style.imagestyle} alt="failed loading the way image." />
          </Link>
        </div>
        <h3 style={style.welcomestyle}>Welcome. Who Are You?</h3>        
        <form noValidate autoComplete="off">
          <TextField
            name="givenName"
            label="given name"
            required={true}
            value={this.state.given_name}
            onChange={this.onTextChange}
            margin="normal"
            style={style.welcomestyle} />
          <TextField
            name="familyName"
            label="family name"
            required={true}
            value={this.state.family_name}
            onChange={this.onTextChange}
            margin="normal"
            style={style.welcomestyle} />
        </form>
        <form noValidate autoComplete="off">
          <TextField
            name="mei"
            label="名"
            required={true}
            value={this.state.mei}
            onChange={this.onTextChange}
            margin="normal"
            style={style.welcomestyle} />
          <TextField
            name="sei"
            label="姓"
            required={true}
            value={this.state.sei}
            onChange={this.onTextChange}
            margin="normal"
            style={style.welcomestyle} />
        </form>
        <div>
          <img src={this.state.iconSrc} style={style.iconstyle}/>
        </div>
        <div>
          <input type="file" accept="image" name="icon" onChange={this.onTextChange}/>
        </div>
        <div>
          <CheckInstax/>
        </div>
        <Button variant="contained"  onClick={this.onSendProfile}>
          GO
        </Button>
      </div>
    );
  }
}

export default withRouter(Signup);
