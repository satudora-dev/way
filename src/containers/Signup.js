import Signup from '../components/Signup';
import { connect } from 'react-redux';

import { signUpAsUser } from '../actions/signup';

const mapStateToProps = state => {
  const currentUserID = state.auth.currentUserID;
  return {
    currentUserID: currentUserID,
    hasOwnProfile: state.users[currentUserID] !== undefined,
  }
}

const mapDispatchToProps = {
  signUpAsUser: signUpAsUser
}

export default connect(mapStateToProps,mapDispatchToProps)(Signup);
