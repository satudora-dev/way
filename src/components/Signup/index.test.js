import React from 'react';
import Signup from '../Signup';
import renderer from 'react-test-renderer';

test('Signup changed by user input', () => {
  const component = renderer.create(
    <Signup />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  // let testEvent = {target: {
  //   name: 'givenName',
  //   value: 'kouhei-Maru'
  // }};
  // tree.props.onChange(testEvent);
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
});
