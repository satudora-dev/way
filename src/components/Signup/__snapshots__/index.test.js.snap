// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup changed by user input 1`] = `
<div
  className="Login"
>
  <div
    style={
      Object {
        "margin-top": "30px",
      }
    }
  >
    <div
      style={
        Object {
          "padding-top": "30px",
          "text-align": "center",
        }
      }
    >
      <img
        alt="loading the way image..."
        src="./way.png"
        style={
          Object {
            "height": 128,
            "vertical-align": "middle",
            "width": 256,
          }
        }
      />
      <h2>
        AI TOKYO LAB & Co.
      </h2>
    </div>
  </div>
  <h3
    style={
      Object {
        "color": "grey",
        "font-weight": "lighter",
      }
    }
  >
    Welcome.
    <span
      style={
        Object {
          "color": "black",
          "font-weight": "bold",
        }
      }
    >
       W
    </span>
    ho
    <span
      style={
        Object {
          "color": "black",
          "font-weight": "bold",
        }
      }
    >
       A
    </span>
    re
    <span
      style={
        Object {
          "color": "black",
          "font-weight": "bold",
        }
      }
    >
       Y
    </span>
    ou?
  </h3>
  <form
    autoComplete="off"
    noValidate={true}
  >
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2"
      style={
        Object {
          "color": "grey",
          "font-weight": "lighter",
        }
      }
    >
      <label
        className="MuiFormLabel-root-10 MuiFormLabel-required-15 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
        data-shrink={false}
      >
        given name
        <span
          className="MuiFormLabel-asterisk-16"
        >
           *
        </span>
      </label>
      <div
        className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
      >
        <input
          aria-invalid={false}
          className="MuiInput-input-25"
          disabled={false}
          name="givenName"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2"
      style={
        Object {
          "color": "grey",
          "font-weight": "lighter",
        }
      }
    >
      <label
        className="MuiFormLabel-root-10 MuiFormLabel-required-15 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
        data-shrink={false}
      >
        family name
        <span
          className="MuiFormLabel-asterisk-16"
        >
           *
        </span>
      </label>
      <div
        className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
      >
        <input
          aria-invalid={false}
          className="MuiInput-input-25"
          disabled={false}
          name="familyName"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
        />
      </div>
    </div>
  </form>
  <form
    autoComplete="off"
    noValidate={true}
  >
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2"
      style={
        Object {
          "color": "grey",
          "font-weight": "lighter",
        }
      }
    >
      <label
        className="MuiFormLabel-root-10 MuiFormLabel-required-15 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
        data-shrink={false}
      >
        名
        <span
          className="MuiFormLabel-asterisk-16"
        >
           *
        </span>
      </label>
      <div
        className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
      >
        <input
          aria-invalid={false}
          className="MuiInput-input-25"
          disabled={false}
          name="mei"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2"
      style={
        Object {
          "color": "grey",
          "font-weight": "lighter",
        }
      }
    >
      <label
        className="MuiFormLabel-root-10 MuiFormLabel-required-15 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
        data-shrink={false}
      >
        姓
        <span
          className="MuiFormLabel-asterisk-16"
        >
           *
        </span>
      </label>
      <div
        className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
      >
        <input
          aria-invalid={false}
          className="MuiInput-input-25"
          disabled={false}
          name="sei"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
    </div>
  </form>
  <div>
    <div
      style={
        Object {
          "margin": "auto",
          "position": "relative",
          "width": "256px",
        }
      }
    >
      <div>
        <img
          src="/portrait.png"
          style={
            Object {
              "borderRadius": "50%",
              "height": "256px",
              "marginTop": "30px",
              "objectFit": "cover",
              "transform": "undefined undefined ",
              "width": "256px",
            }
          }
        />
      </div>
      <div>
        <input
          onChange={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          type="file"
        />
        <button
          className="MuiButtonBase-root-56 MuiButton-root-30 MuiButton-fab-47 MuiButton-mini-52 MuiButton-contained-41 MuiButton-raised-44"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "backgroundColor": "#04B486",
              "color": "white",
              "margin": 0,
              "position": "absolute",
              "right": "32px",
              "top": "240px",
            }
          }
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-31"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-59"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-66"
          />
        </button>
      </div>
    </div>
  </div>
  <button
    className="MuiButtonBase-root-56 MuiButton-root-30 MuiButton-contained-41 MuiButton-raised-44"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-31"
    >
      GO
    </span>
    <span
      className="MuiTouchRipple-root-66"
    />
  </button>
</div>
`;
